From b24a5c2c8db6de7e4323f5b95e83c39b105c5588 Mon Sep 17 00:00:00 2001
From: Michal Mazur <mmazur2@marvell.com>
Date: Thu, 20 Feb 2020 20:36:56 +0100
Subject: [PATCH 0450/1921] gpio: thunderx: Configure pin function at probe

Introduces new device tree option 'pin-cfg' which adds possibility to
setup mode of one or several GPIO pins. It can be used to choose
pin's function, filters, polarity and direction.

Change-Id: Ie5080bab1f683dad8abf57fade5b60cb060e0337
Signed-off-by: Michal Mazur <mmazur2@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/23970
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Radha Chintakuntla <radhac@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26945
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/gpio/gpio-thunderx.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/gpio/gpio-thunderx.c b/drivers/gpio/gpio-thunderx.c
index 2d7feacaf24c..bcf728b7cac6 100644
--- a/drivers/gpio/gpio-thunderx.c
+++ b/drivers/gpio/gpio-thunderx.c
@@ -803,6 +803,31 @@ static int thunderx_gpio_to_irq(struct gpio_chip *chip, unsigned int offset)
 	return irq_find_mapping(txgpio->irqd, offset);
 }
 
+static void thunderx_gpio_pinsel(struct device *dev,
+				 struct thunderx_gpio *txgpio)
+{
+	struct device_node *node;
+	const __be32 *pinsel;
+	int npins, rlen, i;
+	uint32_t pin, sel;
+
+	node = dev_of_node(dev);
+	if (!node)
+		return;
+
+	pinsel = of_get_property(node, "pin-cfg", &rlen);
+	if (!pinsel || rlen % 2)
+		return;
+	npins = rlen / sizeof(__be32) / 2;
+
+	for (i = 0; i < npins; i++) {
+		pin = of_read_number(pinsel++, 1);
+		sel = of_read_number(pinsel++, 1);
+		dev_info(dev, "Set GPIO pin %d CFG register to %x\n", pin, sel);
+		writeq(sel, txgpio->register_base + bit_cfg_reg(pin));
+	}
+}
+
 static int thunderx_gpio_probe(struct pci_dev *pdev,
 			       const struct pci_device_id *id)
 {
@@ -954,6 +979,9 @@ static int thunderx_gpio_probe(struct pci_dev *pdev,
 	dev_info(dev, "ThunderX GPIO: %d lines with base %d.\n",
 		 ngpio, chip->base);
 
+	/* Configure default functions of GPIO pins */
+	thunderx_gpio_pinsel(dev, txgpio);
+
 #ifdef CONFIG_MRVL_OCTEONTX_EL0_INTR
 	if (pdev->irq != 0) {
 		err = gpiochip_irqchip_add(chip, &thunderx_gpio_spi_irq_chip, 0,
-- 
2.31.1

